File: dart/example/advanced_example.dart
//ignore_for_file: unused_import

import 'dart:io' show sleep;
import 'package:quicktrace/quicktrace.dart';

void main() async {
  print('🎯 Advanced QuickTrace Example');
  print('===============================');
  
  // Advanced usage với smart filtering
  final tracer = QuickTracer(
    'Advanced Example',
    outputStyle: OutputStyle.styleDetailed,
    hideUltraFast: Duration(milliseconds: 1),
    showSlowOnly: Duration(milliseconds: 10),
    groupSimilar: Duration(milliseconds: 5),
  );

  // Simulate database operations
  await Future.delayed(Duration(milliseconds: 100));
  tracer.span('Connect to database');

  await Future.delayed(Duration(milliseconds: 45));
  tracer.span('Execute query 1');

  await Future.delayed(Duration(milliseconds: 50)); // Similar to query 1
  tracer.span('Execute query 2');

  await Future.delayed(Duration(milliseconds: 5));
  tracer.span('Cache result');

  await Future.delayed(Duration(microseconds: 500)); // Will be hidden
  tracer.span('Ultra fast operation');

  await Future.delayed(Duration(milliseconds: 200));
  tracer.span('Process business logic');

  await Future.delayed(Duration(milliseconds: 30));
  tracer.span('Send notification');

  tracer.end();
}

File: java/src/main/java/com/leduy/quicktrace/examples/AdvancedExample.java
package com.leduy.quicktrace.examples;

import com.leduy.quicktrace.OutputStyle;
import com.leduy.quicktrace.Tracer;
import java.time.Duration;

/**
 * Advanced usage với smart filtering
 */
public class AdvancedExample {
    
    public static void main(String[] args) throws InterruptedException {
        // Advanced usage với smart filtering
        Tracer tracer = Tracer.builder("Advanced Example")
                .outputStyle(OutputStyle.DETAILED)
                .hideUltraFast(Duration.ofMillis(1))
                .showSlowOnly(Duration.ofMillis(10))
                .groupSimilar(Duration.ofMillis(5))
                .build();
        
        // Simulate database operations
        tracer.span("Connect to database");
        Thread.sleep(100);
        
        tracer.span("Execute query 1");
        Thread.sleep(45);
        
        tracer.span("Execute query 2");
        Thread.sleep(50); // Similar to query 1
        
        tracer.span("Cache result");
        Thread.sleep(5);
        
        tracer.span("Ultra fast operation");
        Thread.sleep(1); // Will be hidden
        
        tracer.span("Process business logic");
        Thread.sleep(200);
        
        tracer.span("Send notification");
        Thread.sleep(30);
        
        tracer.end();
    }
}
